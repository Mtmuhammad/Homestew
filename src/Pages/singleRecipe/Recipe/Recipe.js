import { React, useState, useEffect } from "react";
import Spinner from "../../../Components/Spinner/Spinner";
import client from "../../../http-common";
import "./Recipe.scss";

const Recipe = ({ id }) => {
  const [recipe, setRecipe] = useState({});
  const [isLoading, setIsLoading] = useState(false);
  const [dishTypes, setDishTypes] = useState([]);
  const [ingredients, setIngredients] = useState([]);
  const [directions, setDirections] = useState([]);
  const [tags, setTags] = useState([]);
  const [wines, setWines] = useState([]);
  const [wineText, setWineText] = useState("");
  const [quote, setQuote] = useState([]);

  useEffect(() => {
    const getQuote = () => {
      client
        .get(process.env.REACT_APP_QUOTE_API_URL, {
          headers: {
            "X-Api-Key": process.env.REACT_APP_QUOTE_API_KEY,
          },
        })
        .then(function (res) {
          setQuote(res.data[0]);
        })
        .catch(function (error) {
          console.log(error);
        });
    };
    const getRecipe = () => {
      setIsLoading(true);
      client
        .get(`${id}/information`, {
          params: {
            apiKey: process.env.REACT_APP_API_KEY3,
            addWinePairing: true
          },
        })
        .then(function (res) {
          setRecipe(res.data);
          setDishTypes(res.data.dishTypes);
          setIngredients(res.data.extendedIngredients);
          setDirections(res.data.analyzedInstructions[0].steps);
          setTags(res.data.diets);
          setWines(res.data.winePairing.pairedWines);
          setWineText(res.data.winePairing.pairingText);
          setIsLoading(false);
        })
        .catch(function (error) {
          console.log(error);
        });
    };

    getRecipe();
    getQuote();
  }, [id]);

  return (
    <>
      {isLoading && <Spinner />}
      {recipe && (
        <>
          <div className="col-lg-12">
            <img
              className="img-fluid single-recipe-img"
              src={recipe.image}
              alt={recipe.title}
            />
          </div>

          <div className="col-lg-8 mx-auto">
            <div className="single-recipe-title">
              <div className="single-post-title-text">
                <ul className="ps-0 label">
                  {dishTypes.map((d) => (
                    <li key={d}>{d}</li>
                  ))}
                </ul>
                <h4>{recipe.title}</h4>
                <ul className="ps-0 widget">
                  <li>by {recipe.sourceName || "Anonymous"}</li>
                  <li>{recipe.healthScore}% recommend this recipe</li>
                  <li>
                    <i className="me-2 fa-solid fa-heart"></i>{" "}
                    {recipe.aggregateLikes} Likes{" "}
                  </li>
                </ul>
              </div>
            </div>

            <div className="single-recipe-description">
              <p dangerouslySetInnerHTML={{ __html: recipe.summary }}></p>
            </div>

            <blockquote className="single-recipe-quote">
              <h5>
                <i className="fa-solid fa-quote-left"></i> {quote.quote}{" "}
                <i className="fa-solid fa-quote-right"></i>
              </h5>
              <h6 className="text-muted">{quote.author}</h6>
            </blockquote>

            <div className="single-recipe-description">
              <p dangerouslySetInnerHTML={{ __html: wineText }}></p>
            </div>

            <div className="single-recipe-options text-center">
              <ul>
                <li>
                  <h5>
                    <i className="fa fa-user"></i> SERVES
                  </h5>
                  <span>{recipe.servings} People</span>
                </li>
                <li>
                  <h5>
                    <i className="fa fa-clock-o"></i> PREP TIME
                  </h5>
                  <span>
                    {recipe.preparationMinutes === -1
                      ? `less than ${Math.abs(
                          recipe.preparationMinutes
                        )} minute`
                      : `${recipe.preparationMinutes} minutes`}
                  </span>
                </li>
                <li>
                  <h5>
                    <i className="fa fa-clock-o"></i> Cook TIME
                  </h5>
                  <span>
                    {recipe.readyInMinutes === -1
                      ? `less than ${Math.abs(recipe.readyInMinutes)} minute`
                      : `${recipe.readyInMinutes} minutes`}
                  </span>
                </li>
                <li>
                  <a href="#0" className="print-btn">
                    <i className="fa fa-print"></i> Print
                  </a>
                </li>
              </ul>
            </div>

            <div className="single-recipe-ingredients tables">
              <h5>
                Ingredients <i className="fa-solid fa-utensils"></i>
              </h5>
              <ul className="ps-0">
                {ingredients.map((i) => (
                  <li key={i.original}>{i.original}</li>
                ))}
              </ul>
            </div>
            <div className="single-recipe-directions tables mb-5">
              <h5>
                Directions <i className="fa-solid fa-utensils"></i>
              </h5>
              <ol className="ps-0">
                {directions.map((d) => (
                  <li key={d.number}>{d.step}</li>
                ))}
              </ol>
            </div>
            <div className="single-recipe-tags">
              <ul className="d-flex flex-wrap ps-0">
                {tags && tags.map((t) => <li key={t}>{t}</li>)}
                {wines && wines.map((w) => <li key={w}>{w}</li>)}
              </ul>
            </div>
          </div>
        </>
      )}
    </>
  );
};

export default Recipe;
