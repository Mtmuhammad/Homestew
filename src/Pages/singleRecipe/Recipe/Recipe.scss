@use "../../../Styles/abstracts/mixins" as *;
@use "../../../Styles/abstracts/variables" as *;

.tables {
  padding: 25px 30px 15px;
  border: 1px solid #e1e1e1;
  border-top: none;
  text-align: left;

  & h5 {
    font-size: 2.7rem;
    text-transform: uppercase;
    margin-bottom: 1.5rem;
  }

  & li {
    list-style: none;
    font-size: 1.8rem;
    line-height: 30px;
    position: relative;
  }
}
.single-recipe {
  &-quote {
    padding-left: 30px;
    border-left: 2px solid $primary-green;

    & .fa-solid {
      color: $primary-green;
      font-size: 2.6rem;
    }

    & h5 {
      margin-bottom: 2.5rem;
      font-size: 2rem;
    }

    & h6 {
      font-size: 1.6rem;
      text-transform: uppercase;
      margin-bottom: 8rem;
    }
  }
  &-tags {
    border-bottom: 1px solid #ebebeb;
    padding-bottom: 30px;

    & li {
      color: #292f3b;
      padding: 5px 20px;
      text-transform: uppercase;
      font-size: 1.3rem;
      background-color: #ebebeb;
      display: inline-block;
      margin-bottom: 3px;
      margin-right: 5px;

      &:hover {
        color: #fff;
        background-color: $primary-green;
      }
    }
  }
  &-directions {
    & h5 .fa-utensils {
      color: $tertiary-color;
      font-size: 3.2rem;
      margin-left: 0.5rem;
    }
    & ol {
      list-style: none;
      counter-reset: item;
      padding-left: 0;
    }
    & li {
      counter-increment: item;
      margin-bottom: 20px;
    }
    & li:before {
      margin-right: 10px;
      content: counter(item);
      background: $primary-green;
      border-radius: 100%;
      color: white;
      width: 3rem;
      text-align: center;
      display: inline-block;

      @include respond(sphone) {
        width: 4rem;
      }
    }
  }

  &-ingredients {
    & h5 .fa-utensils {
      color: $secondary-color;
      font-size: 3.2rem;
      margin-left: 0.5rem;
    }
    & li {
      padding-left: 2rem;
      margin-bottom: 10px;

      &:before {
        position: absolute;
        left: 0;
        top: 13px;
        height: 7px;
        width: 7px;
        border-radius: 50%;
        background: $primary-green;
        content: "";

        @include respond(sphone) {
          top: 11px;
          height: 9px;
          width: 9px;
        }
      }
    }
  }

  &-options {
    background: #f5f5f5;
    border: 1px solid #e1e1e1;
    padding: 25px 30px 20px;

    & li:last-child {
      @include respond(sport) {
        margin: 1.5rem auto 0;
      }
    }

    & .fa {
      color: $primary-green;
      margin-right: 5px;
    }

    & h5 {
      font-size: 1.9rem;
      margin-bottom: 0.5rem;
      text-transform: uppercase;
    }

    & span {
      font-size: 1.5rem;
      color: $icon-hover-grey;
      margin-left: 20px;
    }

    & ul {
      display: flex;
      justify-content: space-between;
      align-items: center;
      flex-wrap: wrap;
    }
    & li {
      list-style: none;
    }

    & .print-btn {
      display: inline-block;
      font-size: 1.4rem;
      padding: 8px 25px 6px;
      color: #111;
      text-transform: uppercase;
      font-weight: 700;
      border: 1px solid $primary-green;
    }
  }

  &-description {
    font-size: 1.8rem;
    line-height: 3.5rem;
    margin-bottom: 5rem;

    & b {
      color: $secondary-color;
    }

    & a {
      color: $tertiary-color;
    }
  }
  &-img {
    height: 65rem;

    @include respond(sport) {
      height: 55rem;
    }
    @include respond(phone) {
      height: 50rem;
    }
  }
  &-title {
    background: #fff;
    box-shadow: -1px 5px 5px rgb(20 14 5 / 10%);
    padding: 3rem;
    position: relative;
    margin-top: -10.5rem;
    margin-bottom: 5rem;
    border: 1px solid $light-primary-green;
    border-radius: 10px;

    &-text {
      overflow: hidden;
    }

    & h4 {
      font-size: 3rem;
      line-height: 34px;
      text-transform: uppercase;
    }

    & .label li {
      list-style: none;
      display: inline-block;
      margin-right: 18px;
      color: $primary-green;
      font-size: 1.5rem;
      text-transform: uppercase;
      position: relative;
      font-weight: 700;

      &:last-child {
        margin-right: 0;

        &:after {
          content: "";
        }
      }

      &:after {
        position: absolute;
        right: -14px;
        top: -2px;
        content: "|";
        color: #d7d7d7;
        margin-right: 4px;
      }
    }

    & .widget li {
      list-style: none;
      display: inline-block;
      margin-right: 3rem;
      font-size: 1.5rem;
      color: $icon-hover-grey;
      position: relative;

      & .fa-heart {
        color: rgb(203, 2, 2);
      }

      &:last-child:after {
        content: "";
      }

      &:after {
        position: absolute;
        right: -18px;
        top: -1px;
        content: "|";
        color: #d7d7d7;
      }
    }
  }
}
