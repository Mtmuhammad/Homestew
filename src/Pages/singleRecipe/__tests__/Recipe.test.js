import React from "react";
import { render, screen } from "@testing-library/react";
import Recipe from "../Recipe/Recipe";

describe("smoke and snapshot tests", () => {
  test("should render without crashing", () => {
    render(<Recipe id={663559}/>);
  });

  test("should match snapshot", () => {
    const { asFragment } = render(<Recipe id={663559}/>);
    expect(asFragment()).toMatchSnapshot();
  });
});

test("should render img", ()=> {
   render(<Recipe id={663559}/>)
   const img = screen.getByRole("img")
   expect(img).toBeInTheDocument()
   expect(img).toHaveClass("img-fluid", "single-recipe-img")
})


