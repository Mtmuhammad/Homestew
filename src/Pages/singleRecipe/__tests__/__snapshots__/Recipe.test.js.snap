// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="spinner-container"
  >
    <div
      class="loading-spinner"
    />
  </div>
  <div
    class="col-lg-12"
  >
    <img
      class="img-fluid single-recipe-img"
    />
  </div>
  <div
    class="col-lg-8 mx-auto"
  >
    <div
      class="single-recipe-title"
    >
      <div
        class="single-post-title-text"
      >
        <ul
          class="ps-0 label"
        />
        <h4 />
        <ul
          class="ps-0 widget"
        >
          <li>
            by Anonymous
          </li>
          <li>
            % recommend this recipe
          </li>
          <li>
            <i
              class="me-2 fa-solid fa-heart"
            />
              Likes 
          </li>
        </ul>
      </div>
    </div>
    <div
      class="single-recipe-description"
    >
      <p />
    </div>
    <blockquote
      class="single-recipe-quote"
    >
      <h5>
        <i
          class="fa-solid fa-quote-left"
        />
          
        <i
          class="fa-solid fa-quote-right"
        />
      </h5>
      <h6
        class="text-muted"
      />
    </blockquote>
    <div
      class="single-recipe-description"
    >
      <p />
    </div>
    <div
      class="single-recipe-options text-center"
    >
      <ul>
        <li>
          <h5>
            <i
              class="fa fa-user"
            />
             SERVES
          </h5>
          <span>
             People
          </span>
        </li>
        <li>
          <h5>
            <i
              class="fa fa-clock-o"
            />
             PREP TIME
          </h5>
          <span>
            undefined minutes
          </span>
        </li>
        <li>
          <h5>
            <i
              class="fa fa-clock-o"
            />
             Cook TIME
          </h5>
          <span>
            undefined minutes
          </span>
        </li>
        <li>
          <a
            class="print-btn"
            href="#0"
          >
            <i
              class="fa fa-print"
            />
             Print
          </a>
        </li>
      </ul>
    </div>
    <div
      class="single-recipe-ingredients tables"
    >
      <h5>
        Ingredients 
        <i
          class="fa-solid fa-utensils"
        />
      </h5>
      <ul
        class="ps-0"
      />
    </div>
    <div
      class="single-recipe-directions tables mb-5"
    >
      <h5>
        Directions 
        <i
          class="fa-solid fa-utensils"
        />
      </h5>
      <ol
        class="ps-0"
      />
    </div>
    <div
      class="single-recipe-tags"
    >
      <ul
        class="d-flex flex-wrap ps-0"
      />
    </div>
  </div>
</DocumentFragment>
`;
